src/lib.rs

```diff
@@ -4221,6 +4255,291 @@ fn map_darwin_target_from_rust_to_compiler_architecture(target: &Target) -> &str
+    no_vectorize_loops: bool,
+    no_vectorize_slp: bool,
+    profile_generate: Option<String>,
+    profile_use: Option<String>,
+    control_flow_guard: Option<String>,
+    lto: Option<String>,
+    relocation_model: Option<String>,
+    embed_bitcode: Option<bool>,
+    force_frame_pointers: Option<bool>,
+    link_dead_code: Option<bool>,
+    no_redzone: Option<bool>,
+    soft_float: Option<bool>,
+}
+
+impl RustcCodegenFlags {
+    // Parse flags obtained from CARGO_ENCODED_RUSTFLAGS
+    fn parse(rustflags_env: &str) -> RustcCodegenFlags {
+        fn is_flag_prefix(flag: &str) -> bool {
+            match flag {
+                "-Z" | "-C" | "--codegen" => true,
+                _ => false,
+            }
+        }
+
+        fn join_flag_prefix<'a>(prev: &'a str, curr: &'a str) -> Cow<'a, str> {
+            match prev {
+                "--codegen" | "-C" => Cow::from(format!("-C{}", curr)),
+                "-Z" => Cow::from(format!("-Z{}", curr)),
+                _ => Cow::from(curr),
+            }
+        }
+
+        let mut codegen_flags = RustcCodegenFlags {
+            branch_protection: None,
+            code_model: None,
+            no_vectorize_loops: false,
+            no_vectorize_slp: false,
+            profile_generate: None,
+            profile_use: None,
+            control_flow_guard: None,
+            lto: None,
+            relocation_model: None,
+            embed_bitcode: None,
+            force_frame_pointers: None,
+            link_dead_code: None,
+            no_redzone: None,
+            soft_float: None,
+        };
+
+        let env_flags: Vec<&str> = rustflags_env.split("\u{1f}").collect();
+
+        if !is_flag_prefix(env_flags[0]) {
+            codegen_flags.set_rustc_flag(env_flags[0]);
+        }
+
+        for i in 1..env_flags.len() {
+            let curr = env_flags[i];
+            let prev = env_flags[i - 1];
+
+            // Do not process prefixes on their own
+            if !is_flag_prefix(curr) {
+                // Concat flags preceded by a prefix
+                let rustc_flag = join_flag_prefix(prev, curr);
+                codegen_flags.set_rustc_flag(&rustc_flag);
+            }
+        }
+
+        codegen_flags
+    }
+
+    fn set_rustc_flag(&mut self, flag: &str) {
+        if flag.starts_with("-Z") {
+            self.set_unstable_rustc_flag(flag);
+        } else {
+            self.set_stable_rustc_flag(flag);
+        }
+    }
+
+    fn set_stable_rustc_flag(&mut self, flag: &str) {
+        let (flag, value) = if let Some((flag, value)) = flag.split_once('=') {
+            (flag, Some(value.to_owned()))
+        } else {
+            (flag, None)
+        };
+
+        match flag {
+            "-Ccode-model" => self.code_model = value,
+            "-Cno-vectorize-loops" => self.no_vectorize_loops = true,
+            "-Cno-vectorize-slp" => self.no_vectorize_slp = true,
+            "-Cprofile-generate" => self.profile_generate = value,
```

Comment:
Nit: I think I'd prefer a style more like:
```suggestion
            "-Cprofile-generate" => self.profile_generate = Some(value.expect("-Cprofile-generate must have a value")),
```

That way, we know that a bare `-Cprofile-generate` wasn't specified (yes, this wouldn't be accepted by `rustc`, but we'd enter a wrong state if it was accepted in the future).